py-polars/src/lazyframe/visit.rs

```diff
@@ -106,22 +131,48 @@ impl NodeTraverser {
                 predicate: None,
                 n_rows: None,
             },
-            predicate: None
+            predicate: None,
         };
         let mut lp_arena = self.lp_arena.write().unwrap();
         lp_arena.replace(self.root, ir);
     }
 
-    fn view_current_node(&self) -> PyObject {
-        // Insert Python objects/struct that map to our plan here
-        todo!()
+    fn view_current_node(&self, py: Python<'_>) -> PyResult<PyObject> {
+        self.view_node(py, self.root.0)
+    }
+
+    fn view_node(&self, py: Python<'_>, node: usize) -> PyResult<PyObject> {
+        let lp_arena = self.lp_arena.read().unwrap();
+        let lp_node = lp_arena.get(Node(node));
+        nodes::into_py(py, lp_node)
     }
 
-    fn view_expression(&self, node: usize) -> PyObject {
+    fn view_expression(&self, py: Python<'_>, node: usize) -> PyResult<PyObject> {
         let expr_arena = self.expr_arena.read().unwrap();
-        let _expr = expr_arena.get(Node(node));
-        // Insert Python objects/struct that map to our expr here
-        todo!()
+        let expr = expr_arena.get(Node(node));
+        expr_nodes::into_py(py, expr)
+    }
+
+    fn replace_expressions(&self, expressions: Vec<(usize, PyExpr)>) -> PyResult<Self> {
+        let mut expr_arena = self.expr_arena.read().unwrap().to_owned();
```

Comment:
I am not convinced we should be going via this route.

This clones a whole tree/ subtree. That means that python gets a copy of a `NodeTraverer` that's not connected to the Polars executor anymore, which keeps this instance.

I think the logic of expression rewriting needs to be done at the cudf side if that's needed. In any case the Polars executor will keep a hold of `NodeTraverser` and hook into cudf via the function set by `set_udf`.

Reply:
For now I think I can do this. What I need to be able to do is set up a mapping from arbitrary aggregation expressions to column objects. The way I use this function is to provide a mapping from nodes to `pl.col("a")` pyexprs with names I control in the context. But since this is a restricted remapping (the `PyExpr` in the `Vec` is only for column expressions) I can also do this on our side by setting up an indirection from node ids in our visitor. If I wanted to remap arbitrary expressions it would be more complicated (since I'd need to be able to construct and insert arbitrary expressions into subtrees).

Reply:
Right, so this is mostly to keep this mapping?:

`{original_agg_expr: col(..)}`? 

We could create an expression mapping data structure that's a bit more tailored for this? Then we could also only copy the expressions we use from the arena, which would be much cheaper.

So if you need more complicated expression mappings later, I think we need to go for a different data-structure here.

Reply:
> Right, so this is mostly to keep this mapping?:
> 
> `{original_agg_expr: col(..)}`?

Yes, that's right.

I _think_ I am not likely to want more complicated expression mappings (and if we did it would likely be as more general expression rewriting which I don't want to be doing at this layer anyway).

So I'd be happy to have a simpler interface that just provides an overlay mapping, if that makes sense.

e.g. perhaps:

```
    /// Add some expressions and return their new node ids
    fn add_expressions(&mut self, expressions: Vec<PyExpr>) -> PyResult<Vec<usize>> {
        let mut expr_arena: std::sync::RwLockWriteGuard<'_, Arena<AExpr>> = self.expr_arena.write().unwrap();
        Ok(expressions.iter().map(|e| to_aexpr(e.inner.clone(), &mut expr_arena).0).collect())
    }
```

And then on the traversal side, I know to replace the old node-ids with the new node-ids when I call `view_expression`. Alternately we could add a slot to the `NodeTraverser` that contains an `Option<Vec<Node>>` (or hashmap, but my guess is the dense `Vec` that is the identity for most mappings is actually better performance). So that `view_expression` is then something like:
```
    fn view_expression(&self, py: Python<'_>, node: usize) -> PyResult<PyObject> {
        let expr_arena = self.expr_arena.read().unwrap();
        let n = match &self.expr_mapping {
            Some(mapping) => *mapping.get(node).unwrap(),
            None => Node(node),
        };
        let expr = expr_arena.get(n);
        expr_nodes::into_py(py, expr)
    }
```

WDYT?

Reply:
Yes, I like that more. The traverser gets a bit more tailored to the consumer type, but it probably is better to have the logic as much in the Rust side as possible anyway.

Reply:
Done.